lib/discourse_gamification/leaderboard_cached_view.rb

```diff
@@ -0,0 +1,180 @@
+# frozen_string_literal: true
+
+module ::DiscourseGamification
+  class LeaderboardCachedView
+    attr_reader :leaderboard
+
+    def initialize(leaderboard)
+      @leaderboard = leaderboard
+    end
+
+    def create
+      periods.each { |period| create_mview(period) }
+    end
+
+    def refresh
+      periods.each { |period| refresh_mview(period) }
+    end
+
+    def delete
+      periods.each { |period, _| delete_mview(period) }
+    end
+
+    def scores(period: "all_time")
+      DB.query("SELECT * FROM #{mview_name(period)}")
+    end
+
+    def self.create_all
+      GamificationLeaderboard.find_each { |leaderboard| self.new(leaderboard).create }
+    end
+
+    def self.refresh_all
+      GamificationLeaderboard.find_each { |leaderboard| self.new(leaderboard).refresh }
+    end
+
+    private
+
+    def create_mview(period)
+      name = mview_name(period)
+
+      total_scores_query = <<~SQL
+        WITH leaderboard AS (
+          SELECT * FROM gamification_leaderboards WHERE id = :leaderboard_id
+        ),
+
+        leaderboard_users AS (
+          SELECT
+            users.id,
+            users.name,
+            users.username,
+            users.uploaded_avatar_id
+          FROM
+            users
+          INNER JOIN
+            user_emails ON user_emails.primary = TRUE AND user_emails.user_id = users.id
+          CROSS JOIN
+            leaderboard
+          WHERE NOT
+            (user_emails.email LIKE '%@anonymized.invalid%')
+          AND
+            users.staged = FALSE
+          AND
+            users.id > 0
+          AND
+            (
+              NOT EXISTS(SELECT 1 FROM anonymous_users a WHERE a.user_id = users.id)
+            )
+          AND
+            (
+              (COALESCE(array_length(leaderboard.included_groups_ids, 1), 0) = 0)
```

Comment:
I'll explore that

Reply:
I extracted the group membership checks, created a join based alternative and tested  these against various leaderboard `included_groups_ids`/`excluded_groups_ids` configurations.

I'm sure with more time we could squeeze some gains by tweaking these, but as it stands, the join based query does **mostly** perform better than the existing subquery based approach  but only marginally (**=~ 0.1x**).  It's however significantly less performant (**=~ 4x**) than the subquery approach if the  leaderboard  does not have `included_groups_ids` and `excluded_groups_ids` configured.  This is understandable since it skips the whole subquery dance with with the `(COALESCE(array_length(lb.included_groups_ids, 1), 0) = 0)` and `(COALESCE(array_length(lb.excluded_groups_ids, 1), 0) = 0)` checks.

1. The current version using subqueries:

```sql
 WITH leaderboard AS (
  SELECT * FROM gamification_leaderboards WHERE id = :leadborad_id
)

SELECT
  DISTINCT u.id
FROM users u
CROSS JOIN leaderboard lb
WHERE (
  (COALESCE(array_length(lb.included_groups_ids, 1), 0) = 0)
  OR
  (EXISTS (SELECT 1 FROM group_users AS gu WHERE gu.group_id = ANY(lb.included_groups_ids) AND gu.user_id = u.id))
) AND (
  (COALESCE(array_length(lb.excluded_groups_ids, 1), 0) = 0)
  OR
  (NOT EXISTS (SELECT 1 FROM group_users AS gu WHERE gu.group_id = ANY(lb.excluded_groups_ids) AND gu.user_id = u.id))
);
```

2. Alternative join based query:

```sql
WITH leaderboard AS (
  SELECT * FROM gamification_leaderboards WHERE id = :leaderboard_id
)

SELECT
    DISTINCT u.id
FROM users u
CROSS JOIN leaderboard lb
LEFT JOIN group_users gu_included ON gu_included.group_id = ANY(lb.included_groups_ids) AND gu_included.user_id = u.id
LEFT JOIN group_users gu_excluded ON gu_excluded.group_id = ANY(lb.excluded_groups_ids) AND gu_excluded.user_id = u.id
WHERE (
  (COALESCE(array_length(lb.included_groups_ids, 1), 0) = 0)
  OR
  (gu_included.group_id IS NOT NULL)
) AND (
  (COALESCE(array_length(lb.excluded_groups_ids, 1), 0) = 0)
  OR
  (gu_excluded.group_id IS NULL)
);
```

@xfalcox unless you spot any quick wins with the join query, I say we keep the subqueries for now.

<details>
  <summary>Sample EXPLAIN output</summary>

###  Without  `included_groups_ids` and `excluded_groups_ids` configured

1. Subquery based approach
```
  Unique  (cost=0.42..1696730.19 rows=28575 width=4) (actual time=0.027..51.755 rows=73155 loops=1)
   ->  Nested Loop  (cost=0.42..1696658.76 rows=28575 width=4) (actual time=0.027..39.729 rows=73155 loops=1)
         Join Filter: (((COALESCE(array_length(gamification_leaderboards.included_groups_ids, 1), 0) = 0) OR (SubPlan 1)) AND ((COALESCE(array_length(gamification_leaderboards.excluded_groups_ids, 1), 0) = 0) OR (NOT (SubPlan 2))))
         ->  Index Only Scan using users_pkey on users u  (cost=0.42..4516.79 rows=73153 width=4) (actual time=0.015..15.142 rows=73155 loops=1)
               Heap Fetches: 16116
         ->  Materialize  (cost=0.00..1.05 rows=1 width=38) (actual time=0.000..0.000 rows=1 loops=73155)
               ->  Seq Scan on gamification_leaderboards  (cost=0.00..1.05 rows=1 width=38) (actual time=0.006..0.006 rows=1 loops=1)
                     Filter: (id = 1)
                     Rows Removed by Filter: 1
         SubPlan 1
           ->  Index Only Scan using index_group_users_on_user_id_and_group_id on group_users gu  (cost=0.29..11.55 rows=1 width=0) (never executed)
                 Index Cond: (user_id = u.id)
                 Filter: (group_id = ANY (gamification_leaderboards.included_groups_ids))
                 Heap Fetches: 0
         SubPlan 2
           ->  Index Only Scan using index_group_users_on_user_id_and_group_id on group_users gu_1  (cost=0.29..11.55 rows=1 width=0) (never executed)
                 Index Cond: (user_id = u.id)
                 Filter: (group_id = ANY (gamification_leaderboards.excluded_groups_ids))
                 Heap Fetches: 0
 Planning Time: 0.276 ms
 Execution Time: 54.997 ms
```

1. Join based approach
```
  HashAggregate  (cost=13644.65..13827.53 rows=18288 width=4) (actual time=205.242..218.921 rows=73155 loops=1)
   Group Key: u.id
   Batches: 1  Memory Usage: 6673kB
   ->  Hash Right Join  (cost=10836.42..13598.93 rows=18288 width=4) (actual time=173.287..186.299 rows=73155 loops=1)
         Hash Cond: (gu_included.user_id = u.id)
         Join Filter: (gu_included.group_id = ANY (gamification_leaderboards.included_groups_ids))
         Rows Removed by Join Filter: 81557
         Filter: ((COALESCE(array_length(gamification_leaderboards.included_groups_ids, 1), 0) = 0) OR (gu_included.group_id IS NOT NULL))
         ->  Seq Scan on group_users gu_included  (cost=0.00..1896.28 rows=81528 width=8) (actual time=0.017..11.520 rows=81558 loops=1)
         ->  Hash  (cost=10607.82..10607.82 rows=18288 width=26) (actual time=134.859..134.863 rows=73155 loops=1)
               Buckets: 131072 (originally 32768)  Batches: 1 (originally 1)  Memory Usage: 4525kB
               ->  Hash Right Join  (cost=6163.79..10607.82 rows=18288 width=26) (actual time=103.390..124.097 rows=73155 loops=1)
                     Hash Cond: (gu_excluded.user_id = u.id)
                     Join Filter: (gu_excluded.group_id = ANY (gamification_leaderboards.excluded_groups_ids))
                     Rows Removed by Join Filter: 81557
                     Filter: ((COALESCE(array_length(gamification_leaderboards.excluded_groups_ids, 1), 0) = 0) OR (gu_excluded.group_id IS NULL))
                     ->  Seq Scan on group_users gu_excluded  (cost=0.00..1896.28 rows=81528 width=8) (actual time=0.012..11.891 rows=81558 loops=1)
                     ->  Hash  (cost=5249.37..5249.37 rows=73153 width=42) (actual time=63.094..63.096 rows=73155 loops=1)
                           Buckets: 131072  Batches: 1  Memory Usage: 5454kB
                           ->  Nested Loop  (cost=0.42..5249.37 rows=73153 width=42) (actual time=0.014..23.228 rows=73155 loops=1)
                                 ->  Seq Scan on gamification_leaderboards  (cost=0.00..1.05 rows=1 width=38) (actual time=0.007..0.010 rows=1 loops=1)
                                       Filter: (id = 1)
                                       Rows Removed by Filter: 3
                                 ->  Index Only Scan using users_pkey on users u  (cost=0.42..4516.79 rows=73153 width=4) (actual time=0.005..15.644 rows=73155 loops=1)
                                       Heap Fetches: 16145
 Planning Time: 0.482 ms
 Execution Time: 222.253 ms
```

###  With  `included_groups_ids` 

1. Subquery based approach
```
 Unique  (cost=0.42..1696730.19 rows=28575 width=4) (actual time=0.071..166.987 rows=97 loops=1)
   ->  Nested Loop  (cost=0.42..1696658.76 rows=28575 width=4) (actual time=0.071..166.948 rows=97 loops=1)
         Join Filter: (((COALESCE(array_length(gamification_leaderboards.included_groups_ids, 1), 0) = 0) OR (SubPlan 1)) AND ((COALESCE(array_length(gamification_leaderboards.excluded_groups_ids, 1), 0) = 0) OR (NOT (SubPlan 2))))
         Rows Removed by Join Filter: 73058
         ->  Index Only Scan using users_pkey on users u  (cost=0.42..4516.79 rows=73153 width=4) (actual time=0.018..16.622 rows=73155 loops=1)
               Heap Fetches: 16328
         ->  Materialize  (cost=0.00..1.05 rows=1 width=38) (actual time=0.000..0.000 rows=1 loops=73155)
               ->  Seq Scan on gamification_leaderboards  (cost=0.00..1.05 rows=1 width=38) (actual time=0.007..0.008 rows=1 loops=1)
                     Filter: (id = 2)
                     Rows Removed by Filter: 3
         SubPlan 1
           ->  Index Only Scan using index_group_users_on_user_id_and_group_id on group_users gu  (cost=0.29..11.55 rows=1 width=0) (actual time=0.001..0.001 rows=0 loops=73155)
                 Index Cond: (user_id = u.id)
                 Filter: (group_id = ANY (gamification_leaderboards.included_groups_ids))
                 Rows Removed by Filter: 1
                 Heap Fetches: 63975
         SubPlan 2
           ->  Index Only Scan using index_group_users_on_user_id_and_group_id on group_users gu_1  (cost=0.29..11.55 rows=1 width=0) (never executed)
                 Index Cond: (user_id = u.id)
                 Filter: (group_id = ANY (gamification_leaderboards.excluded_groups_ids))
                 Heap Fetches: 0
 Planning Time: 0.289 ms
 Execution Time: 167.037 ms
```
3. Join based approach
```
 HashAggregate  (cost=13644.65..13827.53 rows=18288 width=4) (actual time=151.925..152.058 rows=97 loops=1)
   Group Key: u.id
   Batches: 1  Memory Usage: 793kB
   ->  Hash Right Join  (cost=10836.42..13598.93 rows=18288 width=4) (actual time=106.943..151.875 rows=97 loops=1)
         Hash Cond: (gu_included.user_id = u.id)
         Join Filter: (gu_included.group_id = ANY (gamification_leaderboards.included_groups_ids))
         Rows Removed by Join Filter: 81460
         Filter: ((COALESCE(array_length(gamification_leaderboards.included_groups_ids, 1), 0) = 0) OR (gu_included.group_id IS NOT NULL))
         Rows Removed by Filter: 73058
         ->  Seq Scan on group_users gu_included  (cost=0.00..1896.28 rows=81528 width=8) (actual time=0.014..11.519 rows=81558 loops=1)
         ->  Hash  (cost=10607.82..10607.82 rows=18288 width=26) (actual time=105.811..105.814 rows=73155 loops=1)
               Buckets: 131072 (originally 32768)  Batches: 1 (originally 1)  Memory Usage: 5382kB
               ->  Hash Right Join  (cost=6163.79..10607.82 rows=18288 width=26) (actual time=76.146..95.212 rows=73155 loops=1)
                     Hash Cond: (gu_excluded.user_id = u.id)
                     Join Filter: (gu_excluded.group_id = ANY (gamification_leaderboards.excluded_groups_ids))
                     Rows Removed by Join Filter: 81557
                     Filter: ((COALESCE(array_length(gamification_leaderboards.excluded_groups_ids, 1), 0) = 0) OR (gu_excluded.group_id IS NULL))
                     ->  Seq Scan on group_users gu_excluded  (cost=0.00..1896.28 rows=81528 width=8) (actual time=0.011..11.328 rows=81558 loops=1)
                     ->  Hash  (cost=5249.37..5249.37 rows=73153 width=42) (actual time=37.278..37.280 rows=73155 loops=1)
                           Buckets: 131072  Batches: 1  Memory Usage: 6311kB
                           ->  Nested Loop  (cost=0.42..5249.37 rows=73153 width=42) (actual time=0.014..22.968 rows=73155 loops=1)
                                 ->  Seq Scan on gamification_leaderboards  (cost=0.00..1.05 rows=1 width=38) (actual time=0.008..0.010 rows=1 loops=1)
                                       Filter: (id = 2)
                                       Rows Removed by Filter: 3
                                 ->  Index Only Scan using users_pkey on users u  (cost=0.42..4516.79 rows=73153 width=4) (actual time=0.005..15.497 rows=73155 loops=1)
                                       Heap Fetches: 16366
 Planning Time: 0.431 ms
 Execution Time: 152.178 ms
```
 </details>